kimchi/src/snarky/folding/instance.rs

```diff
@@ -0,0 +1,300 @@
+                let commitment = sys.compute(loc!(), |_| {
+                    let a: Point<F> = private_input.unwrap().u_i.witness_commitments[i].0[j];
+                    a
+                })?;
+                set.push(commitment);
+            }
+            witness_commitments.push(WitnessCommitments(set))
+        }
+        Ok(Instance {
+            hash1,
+            hash2,
+            witness_commitments,
+        })
+    }
+}
+#[derive(Debug, Clone)]
+struct Challenges<F>(Vec<FullChallenge<F>>);
+
+#[derive(Debug, Clone)]
+pub struct RelaxedInstance<F> {
+    pub hash1: FullChallenge<F>,
+    pub hash2: FullChallenge<F>,
+    pub witness_commitments: Vec<WitnessCommitments<F>>,
+    u: FullChallenge<F>,
+    error_commitment: Point<F>,
+    challenges: Vec<Challenges<F>>,
+}
+
+impl<F: PrimeField> RelaxedInstance<FieldVar<F>> {
+    pub fn absorb_into_sponge(&self, sponge: &mut DuplexState<F>, sys: &mut RunState<F>) {
+        for commitment_set in &self.witness_commitments {
+            for commitment in &commitment_set.0 {
+                sponge.absorb(sys, loc!(), commitment);
+            }
+        }
+        sponge.absorb(sys, loc!(), &self.hash1.0 .0);
+        sponge.absorb(sys, loc!(), &self.hash2.0 .0);
+
+        sponge.absorb(sys, loc!(), &self.u.0 .0);
+        sponge.absorb(sys, loc!(), &self.error_commitment);
+        for set in self.challenges.iter() {
+            for challenge in set.0.iter() {
+                sponge.absorb(sys, loc!(), &challenge.0 .0);
+            }
+        }
+    }
+    pub fn fold(
+        self,
+        sys: &mut RunState<F>,
+        other: Instance<FieldVar<F>>,
+        error_terms: [Point<FieldVar<F>>; 2],
+        base: &FieldVar<F>,
+    ) -> SnarkyResult<Self> {
+        // let mut state = DuplexState::new();
+        let mut challenge_generator = ChallengeGenerator::new(sys, &self, &other, None);
+        let r = challenge_generator.squeeze_challenge(sys, base)?;
+        let hash1 = challenge_linear_combination(self.hash1, SmallChallenge(other.hash1), &r);
+        let hash2 = challenge_linear_combination(self.hash2, SmallChallenge(other.hash2), &r);
+        let witness_commitments = self
+            .witness_commitments
+            .into_iter()
+            .zip(other.witness_commitments.into_iter())
+            .map(|(a, b)| {
+                let set =
+                    a.0.into_iter()
+                        .zip(b.0.into_iter())
+                        .map(|(a, b)| commitment_linear_combination(a, b, &r));
+                WitnessCommitments(set.collect())
+            })
+            .collect();
+        let one = FieldVar::constant(F::one());
+        let u = challenge_linear_combination(self.u, SmallChallenge(one.clone()), &r);
+
+        let rr = r.0.mul(&r.0, None, loc!(), sys)?;
+        let [e1, e2] = error_terms;
+        let e1 = ec_scale(e1, &r);
+        let e2 = ec_scale(e2, &SmallChallenge(rr));
+        let error_commitment = ec_add(e1, e2);
```

Comment:
have to also add the errors from each instance, or just the one from the relaxed instance, given that the other is 0